### PR TITLE
New Engine

### PR DESCRIPTION
**Pull request by remyschwab**

### CODE REVIEW

File: src/otf_collector.cpp [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+                break;
+            default:
+                if(current_state->out1_->c_ > SplitU) split_hits++;
+                update_path(kmer, shift_counts, symbol, agent, current_path, ibf, kmer_cache);
+                if(all_bits_zero(current_path)) break; // Immediately get rid of deadend paths
+                item1 = {current_state->out1_, kmer, current_path, split_hits, shift_counts};
+                queue.push(item1);
+                break;
+        }
+    }
```
Comment (SGSSGene) [inline src/otf_collector.cpp:229]: It is nice to declare the variables at the point, where they are needed:
( I havn't tested this code)

```suggestion
    while(!queue.empty())
    {
        auto front = std::move(queue.front()); // Moving the data out of the queue
        queue.pop(); // removing the remaining data from the queue
        
        State* current_state = front.nfa_state_;
        kmer_t const& kmer = front.kmer_;
        path_t const& current_path = front.path_;
        uint8_t split_hits = front.cycles_;
        uint8_t shift_counts = front.shift_count_;
        
        int symbol = current_state->c_;
        switch(symbol)
        {
            case Match:
                path_matrix.raw_data() |= current_path.raw_data();
                break;
            case SplitU:
                queue.emplace(current_state->out1_, kmer, current_path, split_hits, shift_counts);
                if((current_state->out1_->c_ > SplitU)) queue.back().cycles_++;
                queue.emplace(current_state->out2_, kmer, current_path, split_hits, shift_counts);
                if((current_state->out2_->c_ > SplitU)) queue.back().cycles_++;
                break;
            case SplitP:
                if(split_hits == (ibf.k_-1)) // The k-1 removes redundant homopolymer kmers
                {
                    queue.emplace(current_state->out1_, kmer, current_path, 0, shift_counts);
                    break;
                }
                queue.emplace(current_state->out1_, kmer, current_path, 0, shift_counts);
                queue.emplace(current_state->out2_, kmer, current_path, split_hits, shift_counts);
                break;
            case SplitK:
                if(split_hits == ibf.k_)
                {
                    queue.emplace(current_state->out1_, kmer, current_path, 0, shift_counts);
                    break;
                }
                queue.emplace(current_state->out1_, kmer, current_path, 0, shift_counts);
                queue.emplace(current_state->out2_, kmer, current_path, split_hits, shift_counts);
                break;
            default:
                if(current_state->out1_->c_ > SplitU) split_hits++;
                update_path(kmer, shift_counts, symbol, agent, current_path, ibf, kmer_cache);
                if(all_bits_zero(current_path)) break; // Immediately get rid of deadend paths
                queue.emplace(current_state->out1_, kmer, current_path, split_hits, shift_counts);
                break;
        }
    }
```

File: src/korotkov_nfa.cpp [hunk 2]
```diff
@@ -120,7 +124,13 @@ void nextStep(std::stack<keyState *>& stack, keyState* input)
     default:
             stack.push(input);
             break;
-    case Split:
+    case SplitA:
+            e1 = new keyState{input->qGramFrag_, itptr->out1_, nullptr};
+            e2 = new keyState{input->qGramFrag_, itptr->out2_, nullptr};
+            stack.push(e2);
+            stack.push(e1);
```
Comment (SGSSGene) [inline src/korotkov_nfa.cpp:131]: if you change the SplitB the same way, there is no need for `keyState *e1, *e2`.
```suggestion
            stack.push(new keyState{input->qGramFrag_, itptr->out2_, nullptr});
            stack.push(new keyState{input->qGramFrag_, itptr->out1_, nullptr});
```

**Review by SGSSGene — APPROVED** [review]
I couldn't find any broken code.
In general the code suffers a little over very much old-'C' vibes.

One common thing in C (at least in old C) is to declare a variable at the beginning, and then use it later on.
For C++ I would advice it to declare it as late and narrow as possible.
e.g.:
void f() {
int x;
...
x = 5;
...
}
```
This should be
```
void f() {
...
int x = 5;
}
```

File: src/korotkov_nfa.cpp (hunk 2 above)
Comment (remyschwab) [reply src/korotkov_nfa.cpp:131]: Ah whoops, I actually plan to totally get rid of this module! it's Vincent's...

**Pull request merged**